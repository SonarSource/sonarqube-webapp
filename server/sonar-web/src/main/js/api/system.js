import { getJSON } from '../helpers/request.js';
import $ from 'jquery';

// TODO migrate to fetch()
export function setLogLevel (level) {
  let url = baseUrl + '/api/system/change_log_level';
  return $.post(url, { level });
}

export function getSystemInfo () {
  let url = baseUrl + '/api/system/info';
  return getJSON(url);
}
