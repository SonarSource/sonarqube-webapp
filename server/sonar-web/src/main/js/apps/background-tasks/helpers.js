/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;

function format (int, suffix) {
  return `${int}${suffix}`;
}

export function formatDuration (value) {
  if (!value) {
    return '';
  }
  if (value >= ONE_MINUTE) {
    const minutes = Math.round(value / ONE_MINUTE);
    return format(minutes, 'min');
  } else if (value >= ONE_SECOND) {
    const seconds = Math.round(value / ONE_SECOND);
    return format(seconds, 's');
  } else {
    return format(value, 'ms');
  }
}
