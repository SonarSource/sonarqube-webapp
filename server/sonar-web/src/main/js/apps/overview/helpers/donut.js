import d3 from 'd3';
import React from 'react';

let Sector = React.createClass({
  render() {
    let arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}/>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    let radius = this.props.size / 2;
    let pie = d3.layout.pie()
        .sort(null)
        .value(d => d.value);
    let data = this.props.data;
    let sectors = pie(data).map((d, i) => {
      return <Sector key={i} data={d} fill={data[i].fill} radius={radius} thickness={this.props.thickness}/>;
    });
    return <svg width={this.props.size} height={this.props.size}>
      <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
    </svg>;
  }
});
