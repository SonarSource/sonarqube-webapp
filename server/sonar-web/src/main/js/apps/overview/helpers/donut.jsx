import React from 'react';

const Sector = React.createClass({
  render() {
    const arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}></path>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    const radius = this.props.size / 2;
    const pie = d3.layout.pie().sort(null)
        .value(d => {
          return d.value
        });
    const data = this.props.data;
    const sectors = pie(data).map((d, i) => {
      return <Sector
          key={i}
          data={d}
          fill={data[i].fill}
          radius={radius}
          thickness={this.props.thickness}/>;
    });
    return (
        <svg width={this.props.size} height={this.props.size}>
          <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
        </svg>
    );
  }
});
