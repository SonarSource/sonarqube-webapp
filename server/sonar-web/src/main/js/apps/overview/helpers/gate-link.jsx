import React from 'react';

export default React.createClass({
  render() {
    const url = `${baseUrl}/quality_gates/show/${this.props.gate}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
