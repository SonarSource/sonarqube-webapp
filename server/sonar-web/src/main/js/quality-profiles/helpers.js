/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
(function () {

  Handlebars.registerHelper('profileUrl', function (key) {
    return baseUrl + '/profiles/show?key=' + encodeURIComponent(key);
  });

  Handlebars.registerHelper('exporterUrl', function (profile, exporterKey) {
    var url = baseUrl + '/api/qualityprofiles/export';
    url += '?language=' + encodeURIComponent(profile.language);
    url += '&name=' + encodeURIComponent(profile.name);
    if (exporterKey != null) {
      url += '&exporterKey=' + encodeURIComponent(exporterKey);
    }
    return url;
  });

  Handlebars.registerHelper('severityChangelog', function (severity) {
    var label = '<i class="icon-severity-' + severity.toLowerCase() + '"></i>&nbsp;' + t('severity', severity),
        message = tp('quality_profiles.severity_set_to_x', label);
    return new Handlebars.SafeString(message);
  });

  Handlebars.registerHelper('parameterChangelog', function (value, parameter) {
    if (parameter) {
      return new Handlebars.SafeString(tp('quality_profiles.parameter_set_to_x', value, parameter));
    } else {
      return new Handlebars.SafeString(tp('quality_profiles.changelog.parameter_reset_to_default_value_x', parameter));
    }
  });

})();
