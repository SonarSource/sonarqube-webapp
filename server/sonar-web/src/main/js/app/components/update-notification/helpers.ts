/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { isEmpty } from 'lodash';
import { sortUpgrades } from '../../../components/upgrade/utils';
import { SystemUpgrade } from '../../../types/system';

const MONTH_BEFOR_PREVIOUS_LTS_NOTIFICATION = 6;

type GroupedSystemUpdate = {
  [x: string]: Record<string, SystemUpgrade[]>;
};

export const isPreLTSUpdate = (parsedVersion: number[], latestLTS: string) => {
  const [currentMajor, currentMinor] = parsedVersion;
  const [ltsMajor, ltsMinor] = latestLTS.split('.').map(Number);
  return currentMajor < ltsMajor || (currentMajor === ltsMajor && currentMinor < ltsMinor);
};

export const isPreviousLTSUpdate = (
  parsedVersion: number[],
  latestLTS: string,
  systemUpgrades: GroupedSystemUpdate,
) => {
  const [ltsMajor, ltsMinor] = latestLTS.split('.').map(Number);
  let ltsOlderThan6Month = false;
  const beforeLts = isPreLTSUpdate(parsedVersion, latestLTS);
  if (beforeLts) {
    const allLTS = sortUpgrades(systemUpgrades[ltsMajor][ltsMinor]);
    const ltsReleaseDate = new Date(allLTS[allLTS.length - 1]?.releaseDate ?? '');
    if (isNaN(ltsReleaseDate.getTime())) {
      // We can not parse the LTS date.
      // It is unlikly that this could happen but consider LTS to be old.
      return true;
    }
    ltsOlderThan6Month =
      ltsReleaseDate.setMonth(ltsReleaseDate.getMonth() + MONTH_BEFOR_PREVIOUS_LTS_NOTIFICATION) -
        Date.now() <
      0;
  }
  return ltsOlderThan6Month && beforeLts;
};

export const isMinorUpdate = (parsedVersion: number[], systemUpgrades: GroupedSystemUpdate) => {
  const [currentMajor, currentMinor] = parsedVersion;
  const allMinor = systemUpgrades[currentMajor];
  return Object.keys(allMinor)
    .map(Number)
    .some((minor) => minor > currentMinor);
};

export const isPatchUpdate = (parsedVersion: number[], systemUpgrades: GroupedSystemUpdate) => {
  const [currentMajor, currentMinor, currentPatch] = parsedVersion;
  const allMinor = systemUpgrades[currentMajor];
  const allPatch = sortUpgrades(allMinor[currentMinor] || []);

  if (!isEmpty(allPatch)) {
    const [, , latestPatch] = allPatch[0].version.split('.').map(Number);
    const effectiveCurrentPatch = isNaN(currentPatch) ? 0 : currentPatch;
    const effectiveLatestPatch = isNaN(latestPatch) ? 0 : latestPatch;
    return effectiveCurrentPatch < effectiveLatestPatch;
  }
  return false;
};
