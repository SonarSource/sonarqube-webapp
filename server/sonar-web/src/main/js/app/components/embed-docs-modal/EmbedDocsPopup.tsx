/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Link } from 'react-router';
import { SuggestionLink } from './SuggestionsProvider';
import * as theme from '../../../app/theme';
import BubblePopup, { BubblePopupPosition } from '../../../components/common/BubblePopup';
import DetachIcon from '../../../components/icons-components/DetachIcon';
import { translate } from '../../../helpers/l10n';
import { getBaseUrl } from '../../../helpers/urls';

interface Props {
  onClose: () => void;
  popupPosition?: BubblePopupPosition;
  suggestions: Array<SuggestionLink>;
}

export default class EmbedDocsPopup extends React.PureComponent<Props> {
  static contextTypes = {
    onSonarCloud: PropTypes.bool,
    openOnboardingTutorial: PropTypes.func
  };

  onAnalyzeProjectClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    this.context.openOnboardingTutorial();
  };

  renderTitle(text: string) {
    return <li className="dropdown-header">{text}</li>;
  }

  renderSuggestions() {
    if (this.props.suggestions.length === 0) {
      return null;
    }
    return (
      <>
        {this.renderTitle(translate('embed_docs.suggestion'))}
        {this.props.suggestions.map((suggestion, index) => (
          <li key={index}>
            <Link className="display-flex-center" target="_blank" to={suggestion.link}>
              {suggestion.text}
              <DetachIcon className="spacer-left" fill={theme.gray80} size={12} />
            </Link>
          </li>
        ))}
        <li className="divider" />
      </>
    );
  }

  renderIconLink(link: string, icon: string, text: string) {
    return (
      <a href={link} rel="noopener noreferrer" target="_blank">
        <img
          alt={text}
          className="spacer-right"
          height="18"
          src={`${getBaseUrl()}/images/embed-doc/${icon}`}
          width="18"
        />
        {text}
      </a>
    );
  }

  renderSonarCloudLinks() {
    return (
      <React.Fragment>
        <li className="divider" />
        {this.renderTitle(translate('embed_docs.get_support'))}
        <li>
          {this.renderIconLink(
            'https://about.sonarcloud.io/contact/',
            'sc-icon.svg',
            translate('embed_docs.contact_form')
          )}
        </li>
        {this.renderTitle(translate('embed_docs.stay_connected'))}
        <li>
          {this.renderIconLink('https://about.sonarcloud.io/news/', 'sc-icon.svg', 'Product News')}
        </li>
        <li>
          {this.renderIconLink('https://twitter.com/sonarcloud', 'twitter-icon.svg', 'Twitter')}
        </li>
      </React.Fragment>
    );
  }

  renderSonarQubeLinks() {
    return (
      <React.Fragment>
        <li>
          <a href="#" onClick={this.onAnalyzeProjectClick}>
            {translate('embed_docs.analyze_new_project')}
          </a>
        </li>
        <li className="divider" />
        {this.renderTitle(translate('embed_docs.get_support'))}
        <li>
          {this.renderIconLink(
            'https://groups.google.com/forum/#!forum/sonarqube',
            'google-group-icon.svg',
            'Google Groups'
          )}
        </li>
        <li>
          {this.renderIconLink(
            'http://stackoverflow.com/questions/tagged/sonarqube',
            'so-icon.svg',
            'Stack Overflow'
          )}
        </li>
        {this.renderTitle(translate('embed_docs.stay_connected'))}
        <li>
          {this.renderIconLink('https://blog.sonarsource.com/', 'sq-icon.svg', 'Product News')}
        </li>
        <li>
          {this.renderIconLink('https://twitter.com/SonarQube', 'twitter-icon.svg', 'Twitter')}
        </li>
      </React.Fragment>
    );
  }

  render() {
    return (
      <BubblePopup
        customClass="bubble-popup-bottom bubble-popup-menu abs-width-240 embed-docs-popup"
        position={this.props.popupPosition}>
        <ul className="menu">
          {this.renderSuggestions()}
          <li>
            <Link className="display-flex-center" target="_blank" to="/documentation">
              {translate('embed_docs.documentation_index')}
              <DetachIcon className="spacer-left" fill={theme.gray80} size={12} />
            </Link>
          </li>
          <li>
            <Link onClick={this.props.onClose} to="/web_api">
              {translate('api_documentation.page')}
            </Link>
          </li>
          {this.context.onSonarCloud && this.renderSonarCloudLinks()}
          {!this.context.onSonarCloud && this.renderSonarQubeLinks()}
        </ul>
      </BubblePopup>
    );
  }
}
