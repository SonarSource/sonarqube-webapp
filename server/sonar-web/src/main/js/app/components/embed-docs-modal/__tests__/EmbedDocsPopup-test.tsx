/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import EmbedDocsPopup from '../EmbedDocsPopup';
import { isSonarCloud } from '../../../../helpers/system';

jest.mock('../../../../helpers/system', () => ({ isSonarCloud: jest.fn().mockReturnValue(false) }));

const suggestions = [{ link: '#', text: 'foo' }, { link: '#', text: 'bar' }];

it('should display suggestion links', () => {
  const context = {};
  const wrapper = shallow(
    <EmbedDocsPopup
      currentUser={{ isLoggedIn: true }}
      onClose={jest.fn()}
      suggestions={suggestions}
    />,
    {
      context
    }
  );
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('should display analyze new project link when user has permission', () => {
  const wrapper = shallow(
    <EmbedDocsPopup
      currentUser={{ isLoggedIn: true, permissions: { global: ['provisioning'] } }}
      onClose={jest.fn()}
      suggestions={suggestions}
    />
  );
  expect(wrapper.find('[data-test="analyze-new-project"]').exists()).toBe(true);
});

it('should not display analyze new project link when user does not have permission', () => {
  const wrapper = shallow(
    <EmbedDocsPopup
      currentUser={{ isLoggedIn: true }}
      onClose={jest.fn()}
      suggestions={suggestions}
    />
  );
  expect(wrapper.find('[data-test="analyze-new-project"]').exists()).toBe(false);
});

it('should display correct links for SonarCloud', () => {
  (isSonarCloud as jest.Mock<any>).mockReturnValueOnce(true);
  const context = {};
  const wrapper = shallow(
    <EmbedDocsPopup
      currentUser={{ isLoggedIn: true }}
      onClose={jest.fn()}
      suggestions={suggestions}
    />,
    {
      context
    }
  );
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});
