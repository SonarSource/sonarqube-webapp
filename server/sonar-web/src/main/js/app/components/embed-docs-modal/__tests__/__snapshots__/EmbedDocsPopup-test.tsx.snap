// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display suggestion links 1`] = `
<DropdownOverlay>
  <ul
    className="menu abs-width-240"
  >
    <React.Fragment>
      <li
        className="menu-header"
      >
        embed_docs.suggestion
      </li>
      <li
        key="0"
      >
        <Link
          onClick={[MockFunction]}
          onlyActiveOnIndex={false}
          style={Object {}}
          to="#"
        >
          foo
        </Link>
      </li>
      <li
        key="1"
      >
        <Link
          onClick={[MockFunction]}
          onlyActiveOnIndex={false}
          style={Object {}}
          to="#"
        >
          bar
        </Link>
      </li>
      <li
        className="divider"
      />
    </React.Fragment>
    <li>
      <Link
        onClick={[MockFunction]}
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/documentation"
      >
        embed_docs.documentation
      </Link>
    </li>
    <li>
      <Link
        onClick={[MockFunction]}
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/web_api"
      >
        api_documentation.page
      </Link>
    </li>
    <React.Fragment>
      <li>
        <a
          href="#"
          onClick={[Function]}
        >
          embed_docs.analyze_new_project
        </a>
      </li>
      <li
        className="divider"
      />
      <li>
        <a
          href="https://community.sonarsource.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          embed_docs.get_help
        </a>
      </li>
      <li
        className="divider"
      />
      <li
        className="menu-header"
      >
        embed_docs.stay_connected
      </li>
      <li>
        <a
          href="https://www.sonarsource.com/resources/product-news/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="Product News"
            className="spacer-right"
            height="18"
            src="/images/embed-doc/sq-icon.svg"
            width="18"
          />
          Product News
        </a>
      </li>
      <li>
        <a
          href="https://twitter.com/SonarQube"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="Twitter"
            className="spacer-right"
            height="18"
            src="/images/embed-doc/twitter-icon.svg"
            width="18"
          />
          Twitter
        </a>
      </li>
    </React.Fragment>
  </ul>
</DropdownOverlay>
`;
